packages/components/src/range-control/types.ts

```diff
@@ -0,0 +1,282 @@
+/**
+ * External dependencies
+ */
+import type {
+	CSSProperties,
+	FocusEventHandler,
+	MouseEventHandler,
+	MutableRefObject,
+} from 'react';
+
+/**
+ * Internal dependencies
+ */
+import type { BaseControlProps } from '../base-control/types';
+import type { IconType } from '../icon';
+
+export type NumericProps = {
+	/**
+	 * The minimum `value` allowed.
+	 *
+	 * @default 0
+	 */
+	min?: number;
+	/**
+	 * The maximum `value` allowed.
+	 *
+	 * @default 100
+	 */
+	max?: number;
+	/**
+	 * The current value of the range slider.
+	 */
+	value?: number;
+};
+
+export type RangeStep = number | 'any';
+export type RangeMark = { value: number; label?: string };
+export type RangeMarks = boolean | RangeMark[];
+
+export type MarksProps = NumericProps & {
+	/**
+	 * Disables the `input`, preventing new values from being applied.
+	 *
+	 * @default false
+	 */
+	disabled?: boolean;
+	/**
+	 * Renders a visual representation of `step` ticks. Custom mark indicators
+	 * can be provided by an `Array`.
+	 *
+	 * @default false
+	 */
+	marks?: RangeMarks;
+	/**
+	 * The minimum amount by which `value` changes. It is also a factor in
+	 * validation as `value` must be a multiple of `step` (offset by `min`) to
+	 * be valid. Accepts the special string value `any` that voids the
+	 * validation constraint and overrides both `withInputField` and
+	 * `showTooltip` props to `false`.
+	 *
+	 * @default 1
+	 */
+	step?: RangeStep;
+};
+
+export type RangeMarkProps = {
+	isFilled?: boolean;
+	label?: string;
+	disabled?: boolean;
+	key?: string;
+	style?: CSSProperties;
+};
+
+export type ControlledRangeValue = number | '' | null;
+export type SetControlledRangeValue = ( value?: number | null ) => void;
+
+type RenderTooltipContentCallback = (
+	value?: ControlledRangeValue
+) => string | number | null | undefined;
+
+export type RangeControlProps< P > = Pick<
```

Comment:
Since `P` is only used by optional props should it also be optional?
```suggestion
export type RangeControlProps< P = unknown > = Pick<
```

Reply:
Good point. I also think we could give it a more explicit name, maybe? Something like `IconProps` maybe ?

Reply:
Thanks for the suggestion 👍 

I've updated this in https://github.com/WordPress/gutenberg/pull/40535/commits/93c573e5becb4b0ee669547688b17afaf2361957.